include/boost/beast/core/detail/impl/any_buffers.hpp

```diff
@@ -0,0 +1,83 @@
+//
+// Copyright (c) 2016-2021 Vinnie Falco (vinnie dot falco at gmail dot com)
+// Copyright (c) 2021 Richard Hodges (hodges.r@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/beast
+//
+
+namespace boost {
+namespace beast {
+namespace detail {
+
+template<bool IsConst>
+template<class Buffers, class Alloc>
+void any_buffers<IsConst>::impl<Buffers, Alloc>::destroy() noexcept
+{
+    using allocator_type = typename
+        std::allocator_traits<Alloc>::template
+            rebind_alloc<impl>;
+    auto alloc = allocator_type(a_);
+    using atraits = std::allocator_traits<allocator_type>;
+    atraits::destroy(alloc, this);
+    atraits::deallocate(alloc, this);
+}
+
+// ------------------------------------------------------------
+
+template<bool IsConst>
+template<class Buffers, class Alloc>
+auto
+any_buffers<IsConst>::construct(Buffers const& b, Alloc const& a)
+-> impl_base*
+{
+    using impl_type = impl<Buffers, Alloc>;
+    using allocator_type = typename
+        std::allocator_traits<Alloc>::template
+            rebind_alloc<impl_type>;
+    auto alloc = allocator_type(a);
+    using atraits = std::allocator_traits<allocator_type>;
+
+    std::size_t extra = (std::min)(
+        static_cast<std::size_t>(
+            std::distance(net::buffer_sequence_begin(b),
+                          net::buffer_sequence_end(b))),
+        std::size_t(1)) - 1;
+
+    const auto bytes = sizeof(impl_type) + sizeof(value_type) * extra;
+    const auto units = (bytes + sizeof(impl_type) - 1) / sizeof(impl_type);
+    auto p = atraits::allocate(a, units);
```

Comment:
Still wasting memory here, you want to use an allocator with the smallest type that has the alignment common to both `impl_type` and `value_type`.

Reply:
struggling to see how this would work as if we used a different allocator for the allocation and the construction, the address of the object would strictly need to be stored, as it may not be at the address returned by the allocation.

Reply:
ok, I think I have a solution

Reply:
@glenfe ?

Reply:
Given one `T` and `n` lots of  `U`, you want to allocate storage for `m` lots of `max(alignof(T), alignof(U))`. It should be obvious how to compute `m`. 

The first `sizeof(T)` bytes are for the one `T`, and the offset of the first `U` can always be computed at compile time.

Reply:
For an example, see `<boost/smart_ptr/allocate_shared_array.hpp>` which allocates 1 control block of type `T` followed by `n` lots of a user type `U`   (which makes `allocate_shared<U[]>(n)` possible).